src/plots/cartesian/index.js

```diff
@@ -179,40 +182,75 @@ exports.plot = function(gd, traces, transitionOpts, makeOnCompleteCallback) {
 };
 
 function plotOne(gd, plotinfo, cdSubplot, transitionOpts, makeOnCompleteCallback) {
+    var traceLayerClasses = constants.traceLayerClasses;
     var fullLayout = gd._fullLayout;
     var modules = fullLayout._modules;
-    // list of plot methods to call (this list includes plot methods of 'gone' modules)
-    var plotMethodsToCall = plotinfo.plotMethods || [];
-    // list of plot methods of visible module on this subplot
-    var plotMethods = [];
-    var i, plotMethod;
-
-    for(i = 0; i < modules.length; i++) {
-        var _module = modules[i];
-
-        if(_module.basePlotModule.name === 'cartesian') {
-            plotMethod = _module.plot;
-            Lib.pushUnique(plotMethodsToCall, plotMethod);
-            Lib.pushUnique(plotMethods, plotMethod);
+    var _module, cdModuleAndOthers, cdModule;
+
+    var layerData = [];
+
+    for(var i = 0; i < traceLayerClasses.length; i++) {
+        var className = traceLayerClasses[i];
+
+        for(var j = 0; j < modules.length; j++) {
+            _module = modules[j];
+
+            if(_module.basePlotModule.name === 'cartesian' &&
+                (_module.layerName || _module.name + 'layer') === className
+            ) {
+                var plotMethod = _module.plot;
+
+                // plot all traces of this type on this subplot at once
+                cdModuleAndOthers = getModuleCalcData(cdSubplot, plotMethod);
+                cdModule = cdModuleAndOthers[0];
+                // don't need to search the found traces again - in fact we need to NOT
+                // so that if two modules share the same plotter we don't double-plot
+                cdSubplot = cdModuleAndOthers[1];
+
+                if(cdModule.length) {
+                    layerData.push({
+                        className: className,
+                        plotMethod: plotMethod,
+                        cdModule: cdModule
+                    });
+                }
+                break;
```

Comment:
Does this need to be a 2D loop? Could we just loop over modules, include the layer index in `layerData`, and sort by that at the end? I guess we'd also need to track which layers we've already included, to ensure uniqueness. But still seems faster than looping over all combinations of `className` and `_module`.